File: adapter/src/main/java/io/prizy/adapters/contest/persistence/entity/ContestEntity.java
package io.prizy.adapters.contest.persistence.entity;

import java.time.Instant;
import java.util.Collection;
import java.util.Set;
import java.util.UUID;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.Table;

import io.prizy.domain.contest.model.ContestCategory;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.With;

/**
 * @author Nidhal Dogga
 * @created 4/24/2022 1:14 PM
 */

@With
@Getter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity(name = "Contest")
@Table(name = "contests")
public class ContestEntity {

  @Id
  @GeneratedValue
  private UUID id;
  @Column
  private String name;
  @Column(name = "from_date")
  private Instant fromDate;
  @Column(name = "to_date")
  private Instant toDate;
  @Column
  private String description;
  @Column
  @Enumerated(EnumType.STRING)
  private ContestCategory category;
  @ElementCollection(fetch = FetchType.EAGER)
  @Column(name = "asset_ids")
  private Collection<String> assetIds;
  @Column(name = "cover_asset_id")
  private String coverAssetId;
  @Column
  private Integer cost;
  @Column(name = "facebook_page")
  private String facebookPage;
  @Column(name = "instagram_page")
  private String instagramPage;
  @Column
  private String website;
  @Column(name = "newsletter_subscription")
  private Boolean newsletterSubscription;
  @Column(name = "adult_only")
  private Boolean adultOnly;
  @OneToMany(
    mappedBy = "contest", fetch = FetchType.LAZY, orphanRemoval = true,
    targetEntity = PackEntity.class, cascade = CascadeType.ALL
  )
  private Set<PackEntity> packs;
  @OneToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
  private MerchantEntity merchant;
  @Column(name = "board_id")
  private UUID boardId;

}

File: jpa-adapter/src/main/kotlin/io/prizy/adapters/contest/persistence/entity/ContestEntity.kt
package io.prizy.adapters.contest.persistence.entity

import io.prizy.domain.contest.model.ContestCategory
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode
import org.hibernate.annotations.LazyCollection
import org.hibernate.annotations.LazyCollectionOption
import java.time.Instant
import java.util.UUID
import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.ElementCollection
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.FetchType
import javax.persistence.Id
import javax.persistence.OneToMany
import javax.persistence.OneToOne
import javax.persistence.Table

/**
 *  @author Nidhal Dogga
 *  @created 4/24/2022 1:14 PM
 */

@Entity(name = "Contest")
@Table(name = "contests")
class ContestEntity(
  @Id
  var id: UUID,
  @Column
  var name: String,
  @Column(name = "from_date")
  var fromDate: Instant,
  @Column(name = "to_date")
  var toDate: Instant,
  @Column
  var description: String,
  @Column
  @Enumerated(EnumType.STRING)
  var category: ContestCategory,
  @ElementCollection(fetch = FetchType.EAGER)
  @Column(name = "media_ids")
  var mediaIds: Set<String> = emptySet(),
  @Column(name = "cover_media_id")
  var coverMediaId: String,
  @Column
  var cost: Int,
  @Column(name = "facebook_page")
  var facebookPage: String?,
  @Column(name = "instagram_page")
  var instagramPage: String?,
  @Column
  var website: String?,
  @Column(name = "newsletter_subscription")
  var newsletterSubscription: Boolean,
  @Column(name = "adult_only")
  var adultOnly: Boolean,
  @Fetch(value = FetchMode.SELECT)
  @LazyCollection(LazyCollectionOption.FALSE)
  @OneToMany(mappedBy = "contest", targetEntity = PackEntity::class, fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
  var packs: Set<PackEntity> = emptySet(),
  @OneToOne(fetch = FetchType.EAGER, cascade = [CascadeType.ALL])
  var merchant: MerchantEntity,
  @Column(name = "board_id")
  var boardId: UUID
)
